library/cgroups.go

```
@@ -0,0 +1,105 @@
+	group := fmt.Sprintf("/sys/fs/cgroup/%s", ppid)
+	if err := os.MkdirAll(group, 0644); err != nil {
+		if errors.Is(err, fs.ErrExist) {
+			log.Println("cgroup", group, "already exists")
+		} else {
+			log.Println("failed to create cgroup for", group, ":", err)
+			return err
+		}
+	}
+
+	// Now we can write the actual cgroup limits ...
+	for _, l := range cg.Limits {
+		lf := fmt.Sprintf("%s/%s", group, l.Var)
+		setVar(l.Value, lf)
+	}
+	return nil
+}
+
+func (r *RealCgroup) Run(ctx context.Context, wrapper string, spa *syscall.SysProcAttr, out *SafeBuffer, cleanup func(), cgDir string, c string, args ...string) (*exec.Cmd, *StatusTracker, <-chan bool, error) {
+	args = append([]string{cgDir, c}, args...)
+	log.Println("running process:", wrapper, c, args)
+	cmd := exec.Command(wrapper, args...)
+
+	cmd.Stdout = out
+	cmd.Stderr = out
+
+	// NOTE: experimental - we may need to do this in the wrapper.
+	cmd.SysProcAttr = spa // linux namespace controls
+
+	done := make(chan bool)
+	tracker := &StatusTracker{Status: Unavailable}
+
+	if err := cmd.Start(); err != nil {
+		fmt.Printf("Error running the exec.Command - %s\n", err)
+		cleanup()
+		os.Exit(1)
+	}
+
+	// NOTE: wrapper will add pid to procs file
+
+	go func() {
+		tracker.SetStatus(Running, 0)
+		werr := cmd.Wait()
+		if werr != nil {
+			var ec int
+			if ee, ok := werr.(*exec.ExitError); !ok {
+				ec = ee.ExitCode()
+			}
+			tracker.SetStatus(Error, ec)
+		} else {
+			tracker.SetStatus(Completed, 0)
+		}
+		log.Println("... process ended")
+		done <- true
+		cleanup()
+	}()
+
+	return cmd, tracker, done, nil
+}
+
+func alterController(cmd, fn string) {
+	writeWithFlags(cmd, fn, os.O_WRONLY|os.O_APPEND)
+}
+
+func setVar(cmd, fn string) {
+	writeWithFlags(cmd, fn, os.O_CREATE|os.O_WRONLY)
+}
+
+func writeWithFlags(cmd, fn string, flag int) {
+	file, err := os.OpenFile(fn, flag, 0644)
+	if err != nil {
+		log.Println("unable to open", fn, "for writing :", err)
```

Similar comments here, let's return the error and let the caller decide if this is a fatal error or something they want to try to recover from.